z80count/z80count.py

```diff
@@ -34,6 +37,180 @@
+]
+
+
+def get_program_args():
+    """Get program arguments.
+
+    Main entry point for the config machinery.
+
+    Gathers arguments from the ``DEFAULTS`` structure, a config file
+    and the command line. Returns a ``argparse.Namespace`` object (as
+    returned by ``argparse.Parser.parse_args``), containing the merged
+    options.
+
+    Values specified in the command line have the highest priority,
+    then the options specified in the config file and finally the
+    default values defined by ``DEFAULTS``.
+
+    """
+    config_file = locate_config_file()
+    if config_file:
+        config = load_config_file(config_file, DEFAULTS)
+    else:
+        config = {i.config_name: i.default for i in DEFAULTS}
+
+    args = parse_command_line()
+
+    # merge "args" and "config"
+    for opt in DEFAULTS:
+        if getattr(args, opt.arg_name) is None:
+            setattr(args, opt.arg_name, config[opt.config_name])
+    return args
+
+
+def load_config_file(config_file, schema):
+    parser = configparser.ConfigParser()
+    parser["z80count"] = {i.config_name: i.default for i in schema}
+    try:
+        parser.read(config_file)
+    except configparser.ParsingError:
+        perror("Error parsing config file. Using defaults.")
+
+    section = parser["z80count"]
+    res = {}
+    for opt in schema:
+        v = section.get(opt.config_name)
+        try:
+            v = opt.type(v)
+        except (ValueError, TypeError):
+            perror(
+                "Error parsing config value for '%s'. Using default.",
+                opt.config_name,
+            )
+            v = opt.default
+        res[opt.config_name] = v
+
+    return res
+
+
+def locate_config_file():
+
+    # TODO: check on windows
+
+    z80count_rc = os.environ.get("Z80COUNT_RC")
+    if z80count_rc and os.isfile(z80count_rc):
+        return z80count_rc
+
+    home_dir = os.path.expanduser("~")
+
+    # NOTE: The XDG standard states:
+    #
+    # $XDG_CONFIG_HOME defines the base directory relative to which
+    # user specific configuration files should be stored. If
+    # $XDG_CONFIG_HOME is either not set or empty, a default equal to
+    # $HOME/.config should be used.
+    #
+    # https://specifications.freedesktop.org/basedir-spec/latest/ar01s03.html
+
+    xdg_config_home = os.environ.get("XDG_CONFIG_HOME")
+    if xdg_config_home is None:
+        xdg_config_home = os.path.join(home_dir, ".config")
+
+    candidate = os.path.join(xdg_config_home, "z80count")
+    if os.path.isfile(candidate):
+        return candidate
+
+    candidate = os.path.join(home_dir, ".z80count")
+    if os.path.isfile(candidate):
+        return candidate
+
+    return None
+
+
+def parse_command_line():
+    parser = argparse.ArgumentParser(
+        description='Z80 Cycle Count',
+        epilog="Copyright (C) 2019 Juan J Martinez <jjm@usebox.net>")
+
+    # NOTE: parser arguments *must* provide a default equal to None in
```

Comment:
You could have used as default the value from the configuration file (that would be default if not conf file was provided).